python/snewpy/rate_calculator.py

```diff
@@ -13,11 +13,281 @@
 from astropy import units as u
 from warnings import warn
 from typing import Dict
+from dataclasses import dataclass
+from typing import Callable
+import scipy.stats as st
 
 #various utility methods
 def center(a):
     return 0.5*(a[1:]+a[:-1])
 
+u.kt = u.def_unit('kt',represents=1e6<<u.kg, doc='kilotonne')
+
+class SmearingMatrix:
+    """A class describing energy smearing, i.e. matrix transforming flux from :math:`E_\nu` (true neutrino energy) to :math:`E_{det}` (observed or smeared energy) space""" 
+    def __init__(self, bins_true:u.Quantity, bins_smeared:u.Quantity, matrix:np.ndarray):
+        """
+        Parameters
+        ----------
+        bins_true: Quantity[energy]
+            array of true energy bin limits
+        bins_smeared: Quantity[energy]
+            array of smeared (detected) energy bin limits
+        matrix: 2D numpy array
+            Smearing matrix of shape ``[len(bins_true)-1,len(bins_smeared)-1]``
+        """
+        self.bins_true = bins_true
+        self.bins_smeared = bins_smeared
+        self.matrix = matrix
+        assert matrix.shape==(len(bins_true)-1,len(bins_smeared)-1), \
+        f"Matrix shape {matrix.shape} is inconsistent with (bins_true-1, bins_smeared-1)={len(bins_true)-1,len(bins_smeared)-1}"
+
+    def apply(self, rate:Container)-> Container:
+        """Apply the smearing to the given rate/flux object, 
+        and produce a smeared one
+
+        Parameters
+        ----------
+        rate: Container
+            Input rate/flux object in true energy space.
+            If the rate is integrated over energy (binned), the energy bins must be equal to :attr:`self.bins_true`.
+
+        Returns
+        -------
+        Container
+            The rate/flux in the smeared energy space. 
+            The energy bins of the returned container are equal to :attr:`self.bins_smeared`
+        """
+        if rate.can_integrate('energy'):
+            rate = rate.integrate('energy', self.bins_true)
+        assert np.allclose(rate.energy<<u.GeV, self.bins_true<<u.GeV), \
+        f"rate.energy: {rate.energy<<u.GeV}!={self.bins_true<<u.GeV}"
+        rateS_array = np.dot(rate.array, self.matrix)
+        rateS = Container(rateS_array,
+                          flavor=rate.flavor, 
+                          time=rate.time, 
+                          energy=self.bins_smeared,
+                          integrable_axes=rate._integrable_axes)
+        return rateS
+
+    @classmethod
+    def from_Gaussian(cls, bins_true:u.Quantity, bins_smeared:u.Quantity,
+                      mean:Callable, sigma:Callable=1<<u.keV):
```

Comment:
- Why not give also initial value for mean?
- Why the unit for sigma is in keV here when later you use MeV (line 115)?

Reply:
Mean value is the average energy response `E_vis` for given neutrino energy `E_nu`. And it strongly depends on the process: for IBD `E_vis=E_nu - 784*u.keV`, while for NC on C12 it's just `E_vis = 15.1*u.MeV`

For sigma: 1 keV here is here as a very small default value, so if user doesn't provide the sigma, they get practically no smearing. And the units conversion is done by astropy.units so it works as expected.